src/D2L.CodeStyle.Annotations/Objects/Immutable.cs

```diff
@@ -26,8 +26,25 @@ public sealed class Immutable : ImmutableAttributeBase { }
 		/// but other mutable classes may sub-class it.
 		/// It is always safe to add this annotation because an analyzer will check
 		/// that it is valid.
-		/// </summary>
 		[AttributeUsage( validOn: AttributeTargets.Class )]
 		public sealed class ImmutableBaseClassAttribute : ImmutableAttributeBase { }
+
+		/// <summary>
+		/// If a class, struct or interface is marked with this annotation it
+		/// means that it's type is immutable if all type arguments marked with [OnlyIf]
+		/// are themselves immutable.
+		/// </summary>
+		[AttributeUsage(
+			validOn: AttributeTargets.Class
+			       | AttributeTargets.Interface
+			       | AttributeTargets.Struct
+		)]
+		public sealed class ConditionallyImmutable : ImmutableAttributeBase {
+
+			[AttributeUsage( validOn: AttributeTargets.GenericParameter )]
+			public sealed class OnlyIf : ImmutableAttributeBase { }

```

Comment:
This makes it pretty verbose, but I think it ties them together pretty well

Reply:
`OnlyIf` is a [seemingly unique identifier](https://github.com/search?l=csharp&q=onlyif&type=Repositories) so I wouldn't be opposed to it just being in the namespace.

When we tried for `Objects.Immutable` everyone hit alt+enter in VS and did `using static ...` anyway, so I wonder a bit if that would happen here (and honestly I might be inclined to do it :) ). The downside would be an extra `using static` all over.

Reply:
`Immutable.OnlyIf` might read just as good and would result in an extra using. Just throwing ideas out there.

Reply:
Like you're suggesting the extra using would come from
* thinking "OnlyIf"
* typing `[OnlyIf]`
* hitting Alt+Enter to get it to work

And not necessarily from "hmm, that's too verbose, I want it to be shorter"

Reply:
Yeah, I think it would naturally come from what you described there.

BUT ALSO if I saw that happening I'd be tempted personally to give up and start doing it intentionally (like we did with `[Objects.Immutable]`)